src/transformers/pipelines/token_classification.py

```diff
@@ -198,56 +174,64 @@ def __call__(self, inputs: Union[str, List[str]], **kwargs):
-                    return_special_tokens_mask=True,
-                    return_offsets_mapping=self.tokenizer.is_fast,
-                )
-                if self.tokenizer.is_fast:
-                    offset_mapping = tokens.pop("offset_mapping").cpu().numpy()[0]
-                elif offset_mappings:
-                    offset_mapping = offset_mappings[i]
-                else:
-                    offset_mapping = None
+        return model_inputs
 
-                special_tokens_mask = tokens.pop("special_tokens_mask").cpu().numpy()[0]
+    def forward(self, model_inputs):
+        # Forward
+        special_tokens_mask = model_inputs.pop("special_tokens_mask")
+        offset_mapping = model_inputs.pop("offset_mapping", None)
+        sentence = model_inputs.pop("sentence")
+        if self.framework == "tf":
+            outputs = self.model(model_inputs.data)[0][0].numpy()
+        else:
+            model_inputs = self.ensure_tensor_on_device(**model_inputs)
+            outputs = self.model(**model_inputs)[0][0].cpu().numpy()
+        return {
+            "outputs": outputs,
+            "special_tokens_mask": special_tokens_mask,
+            "offset_mapping": offset_mapping,
+            "sentence": sentence,
+            **model_inputs,
+        }
 
-                # Forward
-                if self.framework == "tf":
-                    entities = self.model(tokens.data)[0][0].numpy()
-                    input_ids = tokens["input_ids"].numpy()[0]
-                else:
-                    with torch.no_grad():
-                        tokens = self.ensure_tensor_on_device(**tokens)
-                        entities = self.model(**tokens)[0][0].cpu().numpy()
-                        input_ids = tokens["input_ids"].cpu().numpy()[0]
-
-            scores = np.exp(entities) / np.exp(entities).sum(-1, keepdims=True)
-            pre_entities = self.gather_pre_entities(sentence, input_ids, scores, offset_mapping, special_tokens_mask)
-            grouped_entities = self.aggregate(pre_entities, self.aggregation_strategy)
-            # Filter anything that is in self.ignore_labels
-            entities = [
-                entity
-                for entity in grouped_entities
-                if entity.get("entity", None) not in self.ignore_labels
-                and entity.get("entity_group", None) not in self.ignore_labels
-            ]
-            answers.append(entities)
-
-        if len(answers) == 1:
-            return answers[0]
-        return answers
+    def postprocess(self, model_outputs):
+        outputs = model_outputs["outputs"]
+        sentence = model_outputs["sentence"]
+        input_ids = model_outputs["input_ids"][0]
+        offset_mapping = model_outputs["offset_mapping"][0] if model_outputs["offset_mapping"] is not None else None
+        special_tokens_mask = model_outputs["special_tokens_mask"][0].numpy()
+
+        scores = np.exp(outputs) / np.exp(outputs).sum(-1, keepdims=True)
```

Comment:
This is not numerically stable and the logsumexp trick should be used here. I think it's defined in another pipeline? @LysandreJik do you remember where I had suggested a softmax in NumPy recently?

Reply:
No issue in changing that on my end. We already have the softmax used in `text_classification` and I think it does not use a logsumexp trick either. Good place to kill 2 birds with 1 stone.

Reply:
Just checked and the function `softmax` I'm mentioning is in the text-classification pipeline file, so re-using it here would be great!

Reply:
Is is using the trick ? (is that just the exp division ? )

Reply:
Yes, the trick is to substract the max before taking the exp, to have smaller numbers.